libs/langchain-community/src/experimental/text_splitters/semantic_chunker.ts

```diff
@@ -0,0 +1,306 @@
+import { BaseDocumentTransformer, Document } from "@langchain/core/documents"
+import { Embeddings } from "@langchain/core/embeddings"
+
+type Sentence = {
+  sentence: string
+  combined_sentence_embedding?: number[]
+  combined_sentence?: string
+  distance_to_next?: number
+}
+
+type SentenceWithCombinedSentence = Sentence & { combined_sentence: string }
+type SentenceWithEmbedding = Sentence & { combined_sentence_embedding: number[]; distance_to_next: number }
+
+function combineSentences(sentences: Sentence[], sentecesToCombine: number = 1) {
+  /**
+   * Combine sentences based on buffer size.
+   *
+   * @param sentences - List of sentences to combine.
+   * @param sentecesToCombine - Number of sentences to combine. Defaults to 1.
+   * @returns List of sentences with combined sentences.
+   */
+
+  for (let i = 0; i < sentences.length; i++) {
+    let combinedSentence = ""
+
+    // Add sentences before the current one
+    for (let j = i - sentecesToCombine; j < i; j++) {
+      if (j >= 0) {
+        combinedSentence += sentences[j].sentence + " "
+      }
+    }
+
+    // Add the current sentence
+    combinedSentence += sentences[i].sentence
+
+    // Add sentences after the current one
+    for (let j = i + 1; j < i + 1 + sentecesToCombine; j++) {
+      if (j < sentences.length) {
+        combinedSentence += " " + sentences[j].sentence
+      }
+    }
+
+    // Store the combined sentence in the current sentence object
+    sentences[i].combined_sentence = combinedSentence
+  }
+
+  return sentences as SentenceWithCombinedSentence[]
+}
+
+function calculateCosineDistances(sentences: SentenceWithEmbedding[]): [number[], SentenceWithEmbedding[]] {
+  /**
+   * Calculate cosine distances between sentences.
+   *
+   * @param sentences - List of sentences to calculate distances for.
+   * @returns Tuple of distances and sentences.
+   */
+
+  const distances: number[] = []
+
+  for (let i = 0; i < sentences.length - 1; i++) {
+    const embeddingCurrent = sentences[i].combined_sentence_embedding
+    const embeddingNext = sentences[i + 1].combined_sentence_embedding
+
+    // Calculate cosine similarity
+    const similarity = cosineSimilarity(embeddingCurrent, embeddingNext)
+
+    // Convert to cosine distance
+    const distance = 1 - similarity
+
+    // Append cosine distance to the list
+    distances.push(distance)
+
+    // Store distance in the dictionary
+    sentences[i].distance_to_next = distance
+  }
+
+  return [distances, sentences]
+}
+
+// Utility function for cosine similarity
+function cosineSimilarity(vecA: number[], vecB: number[]) {
```

Comment:
Let's use the function from core if possible:

```ts
import { cosineSimilarity } from "@langchain/core/math/utils";
```

Reply:
Resolved with https://github.com/nieve-consulting/langchainjs/commit/0e941769550a2e28aa27584c44d955843896ffa2